src/ucp/core/ucp_request.inl

```diff
@@ -697,6 +774,27 @@ ucp_recv_request_set_user_memh(ucp_request_t *req,
     return UCS_OK;
 }
 
+static UCS_F_ALWAYS_INLINE ucs_status_t ucp_request_datatype_iter_init(
+        ucp_request_t *req, ucp_context_h context, void *buffer,
+        size_t iov_count, ucp_datatype_t datatype, size_t contig_length,
+        uint8_t *sg_count, const ucp_request_param_t *param)
+{
+    ucp_datatype_iter_t *dt_iter = &req->send.state.dt_iter;
+    ucs_status_t status;
+
+    UCS_PROFILE_CALL_VOID(ucp_datatype_iter_init, context, (void*)buffer,
+                          iov_count, datatype, contig_length, 1, dt_iter,
+                          sg_count);
+
+    if (!ucp_request_is_user_memh_valid_v2(
+                req, param, buffer, contig_length, iov_count, dt_iter->dt_class,
+                (ucs_memory_type_t)dt_iter->mem_info.type, &status)) {
+        return status;
+    }
+
+    return ucp_datatype_iter_set_memh(dt_iter, param->memh, iov_count);
```

Comment:
maybe add memh param to ucp_datatype_iter_init? it could be NULL if not set

Reply:
I refactored a bit the code so not sure it's relevant after refactoring. (because now `ucp_datatype_iter_set_memh` is called inside `ucp_datatype_iter_init`).